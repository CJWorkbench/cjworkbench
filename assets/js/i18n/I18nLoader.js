import React from 'react'
import { I18nProvider } from '@lingui/react'
import fetchCatalog from './catalogs'

export class I18nLoader extends React.Component {
  state = {
    catalogs: {},
  }
  
  loadCatalog = (language) => {
      this.setState(state => ({
          catalogs: {
            ...state.catalogs,
            [language]: fetchCatalog(language)
          }
      }))
  }
  
  componentDidMount() {
    this.loadCatalog(this.props.language)
  }

  shouldComponentUpdate(nextProps, nextState) {
    const { language } = nextProps
    const { catalogs } = nextState

    if (language !== this.props.language && !catalogs[language]) {
      this.loadCatalog(language)
      return false
    }

    return true
  }

  render () {
    const { children, language } = this.props
    const { catalogs } = this.state

    // Skip rendering when catalog isn't loaded.
    if (!catalogs[language]) {
        return null
    }

    return (
      <I18nProvider language={language} catalogs={catalogs}>
        {children}
      </I18nProvider>
    )
  }
}
