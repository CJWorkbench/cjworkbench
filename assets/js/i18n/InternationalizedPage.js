import React from 'react'
import PropTypes from 'prop-types'
import { en, el } from 'make-plural/plurals'
import { i18n } from '@lingui/core'
import { I18nProvider } from '@lingui/react'
import { currentLocaleId } from './locales'
import Spinner from '../Spinner'

__webpack_public_path__ = window.STATIC_URL + 'bundles/' // eslint-disable-line

async function loadLocale (localeId) {
  const { messages } = await import(
    /* webpackChunkName: "[request]" */
    `@lingui/loader!../../locale/${localeId}/messages.po`
  )
  i18n.loadLocaleData({
    en: { plurals: en },
    el: { plurals: el }
  })
  i18n.load(localeId, messages)
  i18n.activate(localeId)
  return i18n
}

function useActivatedI18n (localeId) {
  const [i18n, setI18n] = React.useState(null)

  React.useEffect(() => {
    loadLocale(currentLocaleId).then(setI18n)
  }, [])

  return i18n
}

export default function InternationalizedPage (props) {
  const { children } = props
  const i18n = useActivatedI18n(currentLocaleId)

  if (i18n === null) {
    return <Spinner />
  }

  return (
    <I18nProvider i18n={i18n} forceRenderOnLocaleChange={false}>
      {children}
    </I18nProvider>
  )
}
InternationalizedPage.propTypes = {
  children: PropTypes.node.isRequired
}
