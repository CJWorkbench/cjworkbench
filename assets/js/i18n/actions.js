const SET_LOCALE = 'SET_LOCALE';

function reduceSetLocale(state, action){
  return {
    ...state,
    locale: action.locale
  };
}

export function setLocaleAction(locale){
  return {
    type: SET_LOCALE,
    locale: locale
  }
}

export const reducerFunctions = {
  [SET_LOCALE]: reduceSetLocale
}

export function localeReducer(state, action){
  if (action.type in reducerFunctions) {
    // Run a registered reducer
    return reducerFunctions[action.type](state, action)
  }
  
  return state;
}
