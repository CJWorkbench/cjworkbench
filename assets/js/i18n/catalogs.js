import enCatalog from '../../locale/en/messages.po'
import elCatalog from '../../locale/el/messages.po'

const catalogs = {
  en: enCatalog,
  el: elCatalog
}

export default function fetchCatalog (language) {
  return catalogs[language]
}
