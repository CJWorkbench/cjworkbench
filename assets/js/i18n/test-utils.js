import { I18nProvider } from '@lingui/react'
import React from 'react'
import { shape, object } from 'prop-types'
import { mount, shallow } from 'enzyme'

/*
 * Below lie utils for testing with i18n.
 * See https://lingui.js.org/guides/testing.html
 */
// Create the I18nProvider to retrieve context for wrapping around.
const intlProvider = new I18nProvider({
  language: 'en',
  catalogs: {
    en: {}
  }
}, {})

const {
  linguiPublisher: {
    i18n: originalI18n
  }
} = intlProvider.getChildContext()

// You customize the i18n object here:
export const i18n = {
  ...originalI18n,
  // provide _ macro, for trying to unwrap default/id or just passing down the key
  _: key => {
    if (typeof key.defaults === 'string') return key.defaults
    if (typeof key.id === 'string') return key.id
    return key
  }
}

/**
 * When using Lingui `withI18n` on components, props.i18n is required.
 */
function nodeWithI18nProp (node) {
  return React.cloneElement(node, { i18n })
}

/*
 * Methods to use
 */
export function shallowWithI18n (node, { context } = {}) {
  return shallow(
    nodeWithI18nProp(node),
    {
      context: Object.assign({}, context, { i18n })
    }
  )
}

export function mountWithI18n (node, { context, childContextTypes } = {}) {
  const newContext = Object.assign({}, context, { linguiPublisher: { i18n } })
  /*
   * I18nProvider sets the linguiPublisher in the context for withI18n to get
   * the i18n object from.
   */
  const newChildContextTypes = Object.assign({},
    {
      linguiPublisher: shape({
        i18n: object.isRequired
      }).isRequired
    },
    childContextTypes
  )
  return mount(
    nodeWithI18nProp(node),
    {
      context: newContext,
      childContextTypes: newChildContextTypes
    }
  )
}
