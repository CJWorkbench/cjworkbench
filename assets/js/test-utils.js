/* globals jest */
import { createStore, applyMiddleware } from 'redux'
import thunk from 'redux-thunk'
import promiseMiddleware from 'redux-promise-middleware'
import errorMiddleware from './error-middleware'
import { workflowReducer } from './workflow-reducer'

// Returns new mock function that returns given json. Used for mocking "get" API calls
export function jsonResponseMock (json) {
  return jest.fn().mockImplementation(() =>
    Promise.resolve(json)
  )
}

// Returns new mock function that gives an OK HTTP response. Use for mocking "set" API calls
export function okResponseMock () {
  return jsonResponseMock(null)
}

// Helper function to return a promise that resolves after all other promise mocks,
// even if they are chained like Promise.resolve().then(...).then(...)
// Technically: this is designed to resolve on the next macrotask
// https://stackoverflow.com/questions/25915634/difference-between-microtask-and-macrotask-within-an-event-loop-context
export function tick () {
  return new Promise(resolve => {
    setTimeout(resolve, 0)
  })
}

export function sleep (ms) {
  return new Promise(resolve => {
    setTimeout(resolve, ms)
  })
}

/**
 * Return `[ notify, promise ]`; `promise` completes when `notify` is called.
 *
 * Usage:
 *
 *     const [ notify, condition ] = createConditionVariable
 *     const done = condition.then(do_something)
 *     await Promise.all([sleep(0.1).then(notify), done])
 */
export function createConditionVariable () {
  let notify
  const promise = new Promise(resolve => { notify = resolve })
  return [notify, promise]
}

/**
 * Mock Redux store with in-production reducers.
 *
 * This is for integration-style tests: they test the reducer _and_ the caller
 * at the same time.
 *
 * Usage:
 *
 *     const api = { undo: jest.fn() }
 *     const store = mockStoreWithReducer({'workflow': ..., ...}, api)
 *     store.dispatch(Actions.undo())
 *     expect(api.undo).toHaveBeenCalled()
 *     expect(store.getState().workflow).toEqual(...)
 */
export function mockStore (initialState, api = {}) {
  const middlewares = [errorMiddleware(), promiseMiddleware, thunk.withExtraArgument(api)]
  const store = createStore(workflowReducer, initialState, applyMiddleware(...middlewares))
  return store
}
