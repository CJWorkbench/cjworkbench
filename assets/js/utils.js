// ---- Utilities ---
import * as Cookies from 'js-cookie'
import { fromByteArray as base64Encode } from 'base64-js'
import { t } from '@lingui/macro'

export function goToUrl (url) {
  window.location.href = url
}

// Current CSRF token
export const csrfToken = Cookies.get('csrftoken')

// Gets the letter coordinate of a column from its index within the column names array
export function idxToLetter (idx) {
  var letters = ''
  var cidx = parseInt(idx)
  cidx += 1
  do {
    cidx -= 1
    letters = String.fromCharCode(cidx % 26 + 65) + letters
    cidx = Math.floor(cidx / 26)
  } while (cidx > 0)
  return letters
}

// Log to Intercom, if installed
export function logUserEvent (name, metadata) {
  if (!window.APP_ID) return

  // If we're in a lesson, drop the event.
  if (window.initState && window.initState.lessonData) return

  window.Intercom('trackEvent', name, metadata)
}

export function timeDifference (start, end, i18n) {
  const ms = new Date(end) - new Date(start)
  const seconds = Math.floor(ms / 1000)
  const minutes = Math.floor(seconds / 60)
  const hours = Math.floor(minutes / 60)
  const days = Math.floor(hours / 24)
  const years = Math.floor(days / 365.25)

  if (years > 0) {
    return i18n._(t('js.util.timeDifference.ago.years')`${years}y ago`)
  } else if (days > 0) {
    return i18n._(t('js.util.timeDifference.ago.days')`${days}d ago`)
  } else if (hours > 0) {
    return i18n._(t('js.util.timeDifference.ago.hours')`${hours}h ago`)
  } else if (minutes > 0) {
    return i18n._(t('js.util.timeDifference.ago.minutes')`${minutes}m ago`)
  } else if (seconds > 0) {
    return i18n._(t('js.util.timeDifference.ago.seconds')`${seconds}s ago`)
  } else {
    return i18n._(t('js.util.timeDifference.now')`just now`)
  }
}

export function escapeHtml (str) {
  str
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&#039;')

  return str
}

/**
 * Scroll `containerEl` vertically by the smallest amount possible to put
 * `el` in view.
 *
 * Set the container CSS `scroll-behavior: smooth` to make this transition
 * look nice.
 *
 * @param el Element we want to focus
 * @param containerEl Element we'll set scrollTop on so it contains el
 * @param marginTop Minimum number of pixels between containerEl.top and el.top
 * @param marginBottom Minimum number of pixels between containerEl.bottom and el.bottom
 */
export function scrollTo (el, containerEl, marginTop, marginBottom) {
  if (marginTop === undefined) marginTop = 10
  if (marginBottom === undefined) marginBottom = 10

  const elRect = el.getBoundingClientRect()
  const containerRect = containerEl.getBoundingClientRect()

  // Calculate dy: how much do we need to add to containerEl.scrollTop to
  // make el go where we want it to go?
  let dy = 0 // Prefer not to scroll
  if (elRect.bottom + marginBottom > containerRect.bottom) {
    // Scroll down if we need to.
    // We do this before up because some els might be taller than containerEl,
    // and in that case we want scroll-up to override scroll-down (because the
    // top of el is more important than the bottom).
    dy += (elRect.bottom + marginBottom - containerRect.bottom)
  }
  if (dy + elRect.top - marginTop < containerRect.top) {
    // Scroll up if we need to.
    dy -= (containerRect.top - (elRect.top - marginTop))
  }

  containerEl.scrollTop += dy
}

/**
 * Build a unique slug with the given prefix.
 *
 * The slug uses randomly-generated bytes: 9 bytes, base64-encoded with
 * '+' and '/' changed to '-' and '_' (to fit a Django "slug" field).
 *
 * Django 'slug' regex: r'^[-a-zA-Z0-9_]+\Z'
 *
 * According to https://en.wikipedia.org/wiki/Birthday_attack, 9 bytes -- 72
 * bits -- means if clients call this 97,000 times, they have a <10^-12 chance
 * of generating a collision within a given Workflow. This should protect even
 * the most click-happy user. (6 bytes would allow 380 before the chance grows
 * higher, and 380 steps is possible.)
 *
 * (Why 9 bytes and not 8? Because base64-encoding gives the same string
 * length. 9 bytes encoded as base64 gives 12 bytes.)
 */
export function generateSlug (prefix) {
  const bytes = new Uint8Array(9)
  window.crypto.getRandomValues(bytes)
  return prefix + (
    base64Encode(bytes)
      .replace(/\+/g, '-')
      .replace(new RegExp('/', 'g'), '_')
  )
}
