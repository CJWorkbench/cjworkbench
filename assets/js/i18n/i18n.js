import { setupI18n } from "@lingui/core"
import { currentLocale } from './locales'
import { fetchCatalog } from './catalogs'

export default const i18n = setupI18n({
   language: currentLocale,
   catalogs: {
       currentLocale: fetchCatalog(currentLocale)
   }
})
