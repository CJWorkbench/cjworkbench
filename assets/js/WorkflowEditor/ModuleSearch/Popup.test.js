/* globals describe, expect, it, jest */
import React from 'react'
import ConnectedPopup, { Popup } from './Popup'
import { Provider } from 'react-redux'
import { mockStore } from '../../test-utils'
import { generateSlug } from '../../utils'
import { mountWithI18n } from '../../i18n/test-utils'

jest.mock('../../utils')

describe('ModuleSearch Popup', () => {
  const ModuleDefaults = {
    idName: 'default',
    name: 'Default',
    description: 'Default',
    category: 'Add data',
    uses_data: true,
    icon: 'url',
    isLessonHighlight: false
  }

  const modules = {
    enigma: {
      idName: 'enigma',
      name: 'Load from Enigma',
      description: 'Enigma description',
      category: 'Add data',
      uses_data: true,
      icon: 'url',
      isLessonHighlight: true
    },
    filter: {
      idName: 'filter',
      name: 'Filter by Text',
      description: 'Text description',
      category: 'Filter',
      uses_data: true,
      icon: 'filter',
      isLessonHighlight: false
    }
  }
  const modulesArray = [modules.enigma, modules.filter]

  const wrapper = (extraProps = {}) => mountWithI18n(
    <Popup
      tabSlug='tab-1'
      index={2}
      isLessonHighlight={false}
      modules={modulesArray}
      onClose={jest.fn()}
      addModule={jest.fn()}
      onUpdate={jest.fn()}
      {...extraProps}
    />
  )

  it('matches snapshot', () => {
    expect(wrapper()).toMatchSnapshot()
  })

  it('finds all suggestions by default', () => {
    const w = wrapper()
    expect(w.text()).toMatch(/Load from Enigma/)
    expect(w.text()).toMatch(/Filter by Text/)
  })

  it('searches by module name', () => {
    const w = wrapper()

    w.find('input[type="search"]').simulate('change', { target: { value: 'a' } })
    expect(w.text()).toMatch(/Load from Enigma/)
    expect(w.text()).not.toMatch(/Filter by Text/)
  })

  it('searches by module description', () => {
    const w = wrapper()
    w.find('input[type="search"]').simulate('change', { target: { value: 'description' } })
    expect(w.text()).toMatch(/Load from Enigma/)
    expect(w.text()).toMatch(/Filter by Text/)
  })

  it('schedules Popper resize when search results change', () => {
    const w = wrapper()
    w.find('input[type="search"]').simulate('change', { target: { value: 'description' } })
    expect(w.prop('onUpdate')).toHaveBeenCalled()
  })

  it('calls onClose on form reset (e.g., clicking button.close)', () => {
    // search field should be empty at start
    const w = wrapper()
    w.find('form').simulate('reset')
    expect(w.prop('onClose')).toHaveBeenCalled()
    expect(w.prop('addModule')).not.toHaveBeenCalled()
  })

  it('calls onCancel on pressing Escape', () => {
    const w = wrapper()
    w.find('input[type="search"]').simulate('keyDown', { keyCode: 27 })
    expect(w.prop('onClose')).toHaveBeenCalled()
    expect(w.prop('addModule')).not.toHaveBeenCalled()
  })

  it('calls addModule and closes on click', () => {
    const w = wrapper()
    w.find('button[data-module-slug="enigma"]').simulate('click')
    expect(w.prop('addModule')).toHaveBeenCalledWith('tab-1', 2, 'enigma')
    expect(w.prop('onClose')).toHaveBeenCalled()
  })

  it('should sort modules by name', () => {
    const w = wrapper({
      modules: [
        { ...ModuleDefaults, idName: 'a', name: 'Z' },
        { ...ModuleDefaults, idName: 'b', name: 'X' },
        { ...ModuleDefaults, idName: 'c', name: 'Y' }
      ]
    })
    expect(w.text()).toMatch(/X.*Y.*Z/)
  })

  it('should show a popover description on hover', () => {
    const w = wrapper()
    w.find('button[data-module-name="Load from Enigma"]').simulate('mouseEnter')
    expect(w.find('SearchResultDescription')).toHaveLength(1)
  })

  it('should highlight search box based on isLessonHighlight', () => {
    const w = wrapper({ isLessonHighlight: true })
    expect(w.find('.module-search-popup.lesson-highlight')).toHaveLength(1)

    const w2 = wrapper({ isLessonHighlight: false })
    expect(w2.find('.module-search-popup.lesson-highlight')).toHaveLength(0)
  })

  it('should highlight lesson-suggested module', () => {
    const w = wrapper()
    expect(w.find('button[data-module-slug="enigma"]').hasClass('lesson-highlight')).toBe(true)
    expect(w.find('button[data-module-slug="filter"]').hasClass('lesson-highlight')).toBe(false)
  })

  describe('connected component', () => {
    const wrapper = (store, extraProps = {}) => mountWithI18n(
      <Provider store={store}>
        <ConnectedPopup
          tabSlug='tab-1'
          index={2}
          onClose={jest.fn()}
          isLastAddButton={false}
          {...extraProps}
        />
      </Provider>
    )

    it('gets modules from the store', () => {
      const store = mockStore({
        modules: {
          a: { id_name: 'a', name: 'AAA', category: 'Analyze', uses_data: true, description: 'A A', icon: 'a' },
          b: { id_name: 'b', name: 'BBB', category: 'Analyze', uses_data: true, description: 'B B', icon: 'b' }
        }
      })
      const w = wrapper(store)
      expect(w.text()).toMatch(/AAA.*BBB/)
    })

    it('hides !uses_data modules', () => {
      const store = mockStore({
        modules: {
          a: { id_name: 'a', name: 'AAA', category: 'Analyze', uses_data: false, description: 'A A', icon: 'a' },
          b: { id_name: 'b', name: 'BBB', category: 'Analyze', uses_data: true, description: 'B B', icon: 'b' }
        }
      })
      const w = wrapper(store)
      expect(w.text()).not.toMatch(/AAA/)
      expect(w.text()).toMatch(/BBB/)
    })

    it('hides deprecated modules', () => {
      const store = mockStore({
        modules: {
          a: { id_name: 'a', name: 'AAA', category: 'Analyze', uses_data: true, description: 'A A', icon: 'a' },
          b: {
            id_name: 'b',
            name: 'BBB',
            category: 'Analyze',
            uses_data: true,
            description: 'B B',
            icon: 'b',
            deprecated: { end_date: '2001-01-01', message: 'Please disappear' }
          }
        }
      })
      const w = wrapper(store)
      expect(w.text()).toMatch(/AAA/)
      expect(w.text()).not.toMatch(/BBB/)
    })

    it('dispatches addModule', () => {
      const api = { addModule: jest.fn(() => new Promise(() => {})) } // never resolves
      generateSlug.mockImplementation(prefix => prefix + 'X')
      const store = mockStore({
        tabs: {
          'tab-1': {
            wf_module_ids: [1, 2]
          }
        },
        wfModules: {
          1: {},
          2: {}
        },
        modules: {
          a: { id_name: 'a', name: 'AAA', category: 'Analyze', uses_data: true, description: 'A A', icon: 'a' },
          b: { id_name: 'b', name: 'BBB', category: 'Analyze', uses_data: true, description: 'B B', icon: 'b' }
        }
      }, api)
      const onClose = jest.fn()
      const w = wrapper(store, { tabSlug: 'tab-1', index: 2, onClose })
      w.find('button[data-module-slug="a"]').simulate('click')
      expect(api.addModule).toHaveBeenCalledWith('tab-1', 'step-X', 'a', 2, {})
      expect(onClose).toHaveBeenCalled()
    })

    describe('in a typical lesson', () => {
      const store = mockStore({
        lessonData: {
          sections: [
            {
              steps: [
                {
                  highlight: [{ type: 'Module', name: 'AAA', index: 0 }],
                  testJs: 'return false'
                }
              ]
            }
          ]
        },
        workflow: {
          tab_slugs: ['tab-1'],
          selected_tab_position: 0
        },
        tabs: {
          'tab-1': { wf_module_ids: [] }
        },
        modules: {
          a: { id_name: 'a', name: 'AAA', category: 'Analyze', uses_data: true, description: 'A A', icon: 'a' },
          b: { id_name: 'b', name: 'BBB', category: 'Analyze', uses_data: true, description: 'B B', icon: 'b' }
        }
      })

      it('lesson-highlights the desired module', () => {
        const w = wrapper(store, { index: 0 })
        expect(w.find('.module-search-result[data-module-slug="a"].lesson-highlight').text()).toEqual('AAA')
      })

      it('lesson-highlights the entire component', () => {
        const w = wrapper(store, { index: 0 })
        expect(w.find('.module-search-popup.lesson-highlight')).toHaveLength(1)
      })

      it('does not lesson-highlight at the wrong index', () => {
        const w = wrapper(store, { index: 2 })
        expect(w.find('.module-search-result.lesson-highlight')).toHaveLength(0)
        expect(w.find('.module-search-popup.lesson-highlight')).toHaveLength(0)
      })
    })
  })
})
