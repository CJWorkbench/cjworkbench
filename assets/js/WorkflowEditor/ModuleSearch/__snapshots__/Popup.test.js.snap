// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModuleSearch Popup matches snapshot 1`] = `
<Popup
  addModule={[MockFunction]}
  i18n={
    Object {
      "_": [Function],
      "_activeLanguageData": Object {
        "plurals": [Function],
      },
      "_activeMessages": Object {},
      "_catalogs": Object {
        "en": Object {
          "languageData": Object {},
          "messages": Object {},
        },
      },
      "_dev": Object {
        "compile": [Function],
        "loadLanguageData": [Function],
      },
      "_language": "en",
      "_locales": undefined,
      "_missing": undefined,
      "plural": [Function],
      "select": [Function],
      "selectOrdinal": [Function],
      "t": [Function],
    }
  }
  index={2}
  isLessonHighlight={false}
  modules={
    Array [
      Object {
        "category": "Filter",
        "description": "Text description",
        "icon": "filter",
        "idName": "filter",
        "isLessonHighlight": false,
        "name": "Filter by Text",
        "uses_data": true,
      },
      Object {
        "category": "Add data",
        "description": "Enigma description",
        "icon": "url",
        "idName": "enigma",
        "isLessonHighlight": true,
        "name": "Load from Enigma",
        "uses_data": true,
      },
    ]
  }
  onClose={[MockFunction]}
  onUpdate={[MockFunction]}
  tabSlug="tab-1"
>
  <div
    className="module-search-popup"
  >
    <WithI18n
      cancel={[MockFunction]}
      onChange={[Function]}
      value=""
    >
      <I18n
        update={true}
        withHash={true}
      >
        <Prompt
          cancel={[MockFunction]}
          i18n={
            Object {
              "_": [Function],
              "_activeLanguageData": Object {
                "plurals": [Function],
              },
              "_activeMessages": Object {},
              "_catalogs": Object {
                "en": Object {
                  "languageData": Object {},
                  "messages": Object {},
                },
              },
              "_dev": Object {
                "compile": [Function],
                "loadLanguageData": [Function],
              },
              "_language": "en",
              "_locales": undefined,
              "_missing": undefined,
              "plural": [Function],
              "select": [Function],
              "selectOrdinal": [Function],
              "t": [Function],
            }
          }
          onChange={[Function]}
          value=""
        >
          <form
            className="module-search-field"
            onReset={[MockFunction]}
            onSubmit={[Function]}
          >
            <input
              autoComplete="off"
              name="moduleQ"
              onChange={[Function]}
              onKeyDown={[Function]}
              placeholder="Searchâ€¦"
              type="search"
              value=""
            />
            <button
              className="reset"
              title="Close Search"
              type="reset"
            >
              <i
                className="icon-close"
              />
            </button>
          </form>
        </Prompt>
      </I18n>
    </WithI18n>
    <SearchResults
      modules={
        Array [
          Object {
            "category": "Filter",
            "description": "Text description",
            "icon": "filter",
            "idName": "filter",
            "isLessonHighlight": false,
            "name": "Filter by Text",
            "uses_data": true,
          },
          Object {
            "category": "Add data",
            "description": "Enigma description",
            "icon": "url",
            "idName": "enigma",
            "isLessonHighlight": true,
            "name": "Load from Enigma",
            "uses_data": true,
          },
        ]
      }
      onClickModule={[Function]}
      search=""
    >
      <ul
        className="module-search-result-groups"
      >
        <ModuleSearchResultGroup
          activeModule={null}
          key="Add data"
          modules={
            Array [
              Object {
                "category": "Add data",
                "description": "Enigma description",
                "icon": "url",
                "idName": "enigma",
                "isLessonHighlight": true,
                "name": "Load from Enigma",
                "uses_data": true,
              },
            ]
          }
          name="Add data"
          onClickModule={[Function]}
          onMouseEnterModule={[Function]}
        >
          <li
            className="module-search-result-group"
            data-name="Add data"
          >
            <h4>
              Add data
            </h4>
            <ul
              className="module-search-results"
            >
              <SearchResult
                category="Add data"
                description="Enigma description"
                icon="url"
                idName="enigma"
                isActive={false}
                isLessonHighlight={true}
                key="Load from Enigma"
                name="Load from Enigma"
                onClick={[Function]}
                onMouseEnter={[Function]}
                uses_data={true}
              >
                <Manager>
                  <li>
                    <Reference>
                      <InnerReference
                        setReferenceNode={[Function]}
                      >
                        <button
                          className="module-search-result lesson-highlight"
                          data-module-name="Load from Enigma"
                          data-module-slug="enigma"
                          id="module-search-result-enigma"
                          onClick={[Function]}
                          onMouseEnter={[Function]}
                        >
                          <i
                            className="icon-url"
                          />
                          <span
                            className="name"
                          >
                            Load from Enigma
                          </span>
                        </button>
                      </InnerReference>
                    </Reference>
                  </li>
                </Manager>
              </SearchResult>
            </ul>
          </li>
        </ModuleSearchResultGroup>
        <ModuleSearchResultGroup
          activeModule={null}
          key="Filter"
          modules={
            Array [
              Object {
                "category": "Filter",
                "description": "Text description",
                "icon": "filter",
                "idName": "filter",
                "isLessonHighlight": false,
                "name": "Filter by Text",
                "uses_data": true,
              },
            ]
          }
          name="Filter"
          onClickModule={[Function]}
          onMouseEnterModule={[Function]}
        >
          <li
            className="module-search-result-group"
            data-name="Filter"
          >
            <h4>
              Filter
            </h4>
            <ul
              className="module-search-results"
            >
              <SearchResult
                category="Filter"
                description="Text description"
                icon="filter"
                idName="filter"
                isActive={false}
                isLessonHighlight={false}
                key="Filter by Text"
                name="Filter by Text"
                onClick={[Function]}
                onMouseEnter={[Function]}
                uses_data={true}
              >
                <Manager>
                  <li>
                    <Reference>
                      <InnerReference
                        setReferenceNode={[Function]}
                      >
                        <button
                          className="module-search-result"
                          data-module-name="Filter by Text"
                          data-module-slug="filter"
                          id="module-search-result-filter"
                          onClick={[Function]}
                          onMouseEnter={[Function]}
                        >
                          <i
                            className="icon-filter"
                          />
                          <span
                            className="name"
                          >
                            Filter by Text
                          </span>
                        </button>
                      </InnerReference>
                    </Reference>
                  </li>
                </Manager>
              </SearchResult>
            </ul>
          </li>
        </ModuleSearchResultGroup>
      </ul>
    </SearchResults>
  </div>
</Popup>
`;
